File: wrappers/python/nntile/optimizer/adam.py
# @copyright (c) 2022-present Skolkovo Institute of Science and Technology
#                              (Skoltech), Russia. All rights reserved.
#                2023-present Artificial Intelligence Research Institute
#                              (AIRI), Russia. All rights reserved.
#
# NNTile is software framework for fast training of big neural networks on
# distributed-memory heterogeneous systems based on StarPU runtime system.
#
# @file wrappers/python/nntile/optimizer/adam.py
# Implementation of Adam with amsgrad option within nntile package
#
# @version 1.1.0

import pickle

import numpy as np
import torch

import nntile
from nntile.tensor import TensorTraits


class Adam:
    def __init__(
        self,
        params,
        lr,
        next_tag,
        beta1=0.9,
        beta2=0.999,
        weight_decay=0.0,
        eps=1e-8,
        dtype=np.float32,
        start_lr=None,
        full_lr_iter=None,
    ):
        self.params = params
        self.next_tag = next_tag
        self.num_iter = 1
        self.dtype = dtype
        self.first_moments = []
        self.second_moments = []
        for p in self.params:
            p_traits = TensorTraits(p.value.shape, p.value.basetile_shape)
            self.first_moments.append(
                type(p.value)(p_traits, p.value.distribution, self.next_tag)
            )
            self.next_tag = self.first_moments[-1].next_tag
            self.second_moments.append(
                type(p.value)(p_traits, p.value.distribution, self.next_tag)
            )
            self.next_tag = self.second_moments[-1].next_tag
        self.lr = lr
        self.start_lr = start_lr
        self.full_lr_iter = full_lr_iter
        self.beta1 = beta1
        self.beta2 = beta2
        self.weight_decay = weight_decay
        self.eps = eps

    def get_next_tag(self):
        return self.next_tag

    def unregister(self):
        for i in range(len(self.first_moments)):
            self.first_moments[i].unregister()
            self.second_moments[i].unregister()

    def step(self):
        cur_lr = self.lr
        if self.start_lr is not None and self.full_lr_iter is not None:
            if self.num_iter < self.full_lr_iter and self.full_lr_iter > 1:
                cur_lr = (self.lr - self.start_lr) / (self.full_lr_iter - 1)
                cur_lr = cur_lr * (self.num_iter - 1) + self.start_lr
        for i, p in enumerate(self.params):
            nntile.tensor.fused_adam_step(
                p.value,
                p.grad,
                self.first_moments[i],
                self.second_moments[i],
                cur_lr,
                self.eps,
                self.beta1,
                self.beta2,
                self.weight_decay,
                self.num_iter,
            )
            p.value.wont_use()
            # dP can be deleted
            # p.grad.wont_use()
            p.grad.invalidate_submit()
            self.first_moments[i].wont_use()
            self.second_moments[i].wont_use()
        self.num_iter += 1

    def save_state(self, path, dtype="fp32"):
        first_moments = []
        second_moments = []
        for i in range(len(self.first_moments)):
            f_m = np.array(
                np.zeros(self.first_moments[i].shape, dtype=self.dtype),
                order="F",
            )
            self.first_moments[i].to_array(f_m)
            s_m = np.array(
                np.zeros(self.second_moments[i].shape, dtype=self.dtype),
                order="F",
            )
            self.second_moments[i].to_array(s_m)
            if dtype == "fp32":
                first_moments.append(f_m.copy())
                second_moments.append(s_m.copy())
            elif dtype == "fp16":
                first_moments.append(torch.tensor(f_m, dtype=torch.float16))
                second_moments.append(torch.tensor(s_m, dtype=torch.float16))
            elif dtype == "bf16":
                first_moments.append(torch.tensor(f_m, dtype=torch.bfloat16))
                second_moments.append(torch.tensor(s_m, dtype=torch.bfloat16))

        stored_data = {
            "first_moments": first_moments,
            "second_moments": second_moments,
            "num_iter": self.num_iter,
            "beta1": self.beta1,
            "beta2": self.beta2,
            "lr": self.lr,
            "start_lr": self.start_lr,
            "full_lr_iter": self.full_lr_iter,
            "eps": self.eps,
            "weight_decay": self.weight_decay,
        }
        with open(path, "wb") as fp:
            pickle.dump(stored_data, fp)

    def load_state(self, path):
        with open(path, "rb") as fp:
            stored_states = pickle.load(fp)

        self.lr = stored_states["lr"]
        self.start_lr = stored_states["start_lr"]
        self.full_lr_iter = stored_states["full_lr_iter"]
        self.beta1 = stored_states["beta1"]
        self.beta2 = stored_states["beta2"]
        self.eps = stored_states["eps"]
        self.num_iter = stored_states["num_iter"]
        self.weight_decay = stored_states["weight_decay"]

        first_moments = stored_states["first_moments"]
        second_moments = stored_states["second_moments"]
        for i in range(len(first_moments)):
            self.first_moments[i].from_array(
                first_moments[i].to(torch.float32)
            )
            self.second_moments[i].from_array(
                second_moments[i].to(torch.float32)
            )

    def get_nbytes(self):
        nbytes = 0
        for i in range(len(self.first_moments)):
            nbytes += self.first_moments[i].get_nbytes()
            nbytes += self.second_moments[i].get_nbytes()
        return nbytes

File: src/graph/optim/adam.cc
/*! @copyright (c) 2022-present Skolkovo Institute of Science and Technology
 *                              (Skoltech), Russia. All rights reserved.
 *                 2023-present Artificial Intelligence Research Institute
 *                              (AIRI), Russia. All rights reserved.
 *
 * NNTile is software framework for fast training of big neural networks on
 * distributed-memory heterogeneous systems based on StarPU runtime system.
 *
 * @file src/graph/optim/adam.cc
 * Adam optimizer implementation.
 *
 * @version 1.1.0
 * */

#include "nntile/graph/optim/adam.hh"

#include <fstream>
#include <stdexcept>

#include <nlohmann/json.hpp>

#include "nntile/graph/nn/adam_step.hh"

namespace nntile::graph::optim
{

Adam::Adam(NNGraph* graph,
           module::Module* module,
           Scalar lr,
           Scalar beta_1,
           Scalar beta_2,
           Scalar eps,
           Scalar weight_decay)
    : Optimizer(graph, module)
    , lr_(lr)
    , beta_1_(beta_1)
    , beta_2_(beta_2)
    , eps_(eps)
    , weight_decay_(weight_decay)
{
}

void Adam::step()
{
    for(auto& ps : param_states_)
    {
        std::string m1_name = ps.name + "_first_moment";
        std::string m2_name = ps.name + "_second_moment";

        auto* first_moment = graph_->tensor(
            ps.param->shape(), m1_name,
            ps.param->dtype(), false);
        first_moment->mark_input(true);
        first_moment->mark_output(true);

        auto* second_moment = graph_->tensor(
            ps.param->shape(), m2_name,
            ps.param->dtype(), false);
        second_moment->mark_input(true);
        second_moment->mark_output(true);

        ps.param->mark_input(true);
        ps.param->mark_output(true);

        adam_step(ps.param, ps.grad, first_moment, second_moment,
                  num_iter_ + 1, beta_1_, beta_2_, eps_, lr_, weight_decay_);

        ps.buffers.emplace_back(m1_name, first_moment);
        ps.buffers.emplace_back(m2_name, second_moment);
    }
}

void Adam::save_config(const std::string& path) const
{
    nlohmann::json j;
    j["optimizer"] = "Adam";
    j["num_iter"] = num_iter_;
    j["lr"] = lr_;
    j["beta_1"] = beta_1_;
    j["beta_2"] = beta_2_;
    j["eps"] = eps_;
    j["weight_decay"] = weight_decay_;

    nlohmann::json params = nlohmann::json::array();
    for(const auto& ps : param_states_)
    {
        params.push_back(ps.name);
    }
    j["param_names"] = params;

    std::ofstream f(path);
    if(!f.is_open())
    {
        throw std::runtime_error(
            "Adam::save_config: cannot open '" + path + "'");
    }
    f << j.dump(2);
}

void Adam::load_config(const std::string& path)
{
    std::ifstream f(path);
    if(!f.is_open())
    {
        throw std::runtime_error(
            "Adam::load_config: cannot open '" + path + "'");
    }
    nlohmann::json j = nlohmann::json::parse(f);
    if(j.at("optimizer").get<std::string>() != "Adam")
    {
        throw std::runtime_error(
            "Adam::load_config: optimizer type mismatch");
    }
    num_iter_ = j.at("num_iter").get<Index>();
    lr_ = j.at("lr").get<Scalar>();
    beta_1_ = j.at("beta_1").get<Scalar>();
    beta_2_ = j.at("beta_2").get<Scalar>();
    eps_ = j.at("eps").get<Scalar>();
    weight_decay_ = j.at("weight_decay").get<Scalar>();
}

std::string Adam::repr() const
{
    return "Adam(lr=" + std::to_string(lr_) +
           ", beta_1=" + std::to_string(beta_1_) +
           ", beta_2=" + std::to_string(beta_2_) +
           ", eps=" + std::to_string(eps_) +
           ", weight_decay=" + std::to_string(weight_decay_) +
           ", num_params=" + std::to_string(param_states_.size()) + ")";
}

} // namespace nntile::graph::optim
